internal/util/metrics/metrics.go

```diff
@@ -0,0 +1,98 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+type Server struct {
+	pipelinesGauge          *prometheus.GaugeVec
+	pipelineCounterMap      map[string]struct{}
+	isbsvcGauge             *prometheus.GaugeVec
+	isbsvcCounterMap        map[string]struct{}
+	numaflowControllerGauge *prometheus.GaugeVec
+}
+
+var (
+	// the pipelineMap contains the information of all running pipeline with key as "name_namespace"
+	pipelinesGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "numaflow_pipelines_running",
+			Help: "Number of Numaflow pipelines running",
+		},
+		[]string{},
+	)
+	isbsvcGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "numaflow_isb_service_running",
+			Help: "Number of Numaflow ISB Service running",
+		},
+		[]string{},
+	)
+	numaflowControllerGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "numaflow_controller_running",
+			Help: "Number of Numaflow controller running",
+		},
+		[]string{"name", "namespace", "version"},
+	)
+)
+
+func NewMetricsServer() *Server {
+	// Register custom metrics with the global prometheus registry
+	metrics.Registry.MustRegister(pipelinesGauge)
```

Comment:
What is the reason of not using `promauto`?

Reply:
Kube builder have given here https://book.kubebuilder.io/reference/metrics.html?highlight=prometheus#publishing-additional-metrics an example to register custom metrics and have not used `promauto`, so following the same. Do  we have any benefit's over it?

Reply:
I have defined the `Inc` and `Dec` method so that I can update the `map` used to contain pipeline info but both fn is using same `m.pipelinesGauge.WithLabelValues().Set()` operation at the end.

Also `promauto.NewGaugeVec()` is using the same `prometheus.NewGaugeVec()` underlying method just that with `promauto` I don't have to register the metrics manually i.e. `metrics.Registry.MustRegister()`

Reply:
Just want to double confirm this built in metrics server will be running as a separate process from the manager, right?

Reply:
I can see [here](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/manager/internal.go#L408) it's starting as a sub-process but didn't find any reference in the docs.